### PR TITLE
Fix: Implement get_quest_users admin route

### LINKED ISSUE #326: Create New Admin Route get_quest_users
## Description 📹

Implement a new admin route `get_quest_users` that retrieves a list of addresses who have completed **at least one task** of a specific quest. The route will accept a `quest_id` parameter, fetch all task IDs associated with that quest, and then identify the addresses that have completed any of these tasks. This will allow admin users to see which participants have engaged with the quest to any extent. The route should be secured using the existing authentication middleware to ensure only authorized admin users can access this information.

## Proposed Actions 🛠️

Here’s a checklist of actions to follow for resolving this issue:

1. **Fork and Create Branch**:  
   Fork the repository and create a new branch using the issue number:

   ```bash
   git checkout -b fix-[issue-number]
   ```

2. **Implement Changes**:

   - **Create the Route**: add a new handler in `src/endpoints/admin/quest_boost`: `get_quest_users.rs`

   - **Define Request Parameters**:

     ```rust
     pub_struct!(Deserialize; GetQuestUsersParams {
         quest_id: i64,
     });
     ```

   - **Implement the Handler Logic (should look something like that)**:

     ```rust
     #[route(get, "/admin/quests/get_quest_users", auth_middleware)]
     pub async fn get_quest_users_handler(
         State(state): State<Arc<AppState>>,
         Extension(_sub): Extension<String>, // Authentication handled by middleware
         Query(params): Query<GetQuestUsersParams>,
     ) -> impl IntoResponse {
         let tasks_collection = state.db.collection::<QuestTaskDocument>("tasks");
         let completed_tasks_collection = state.db.collection::<CompletedTaskDocument>("completed_tasks");

         // Fetch all task IDs for the given quest_id
         let task_filter = doc! { "quest_id": params.quest_id };
         let task_cursor = match tasks_collection.find(task_filter, None).await {
             Ok(cursor) => cursor,
             Err(e) => return get_error(format!("Error fetching tasks: {}", e)),
         };

         let task_ids_result: Result<Vec<i64>, _> = task_cursor
             .map_ok(|doc| doc.id)
             .collect()
             .await;

         let task_ids = match task_ids_result {
             Ok(ids) => ids,
             Err(e) => return get_error(format!("Error processing tasks: {}", e)),
         };

         if task_ids.is_empty() {
             return get_error(format!("No tasks found for quest_id {}", params.quest_id));
         }

         // Fetch all completed tasks for these task IDs
         let completed_task_filter = doc! { "task_id": { "$in": &task_ids } };
         let completed_task_cursor = match completed_tasks_collection.find(completed_task_filter, None).await {
             Ok(cursor) => cursor,
             Err(e) => return get_error(format!("Error fetching completed tasks: {}", e)),
         };

         let completed_tasks_result: Result<Vec<CompletedTaskDocument>, _> = completed_task_cursor.try_collect().await;

         let users = match completed_tasks_result {
             Ok(completed_tasks) => {
                 let user_set: HashSet<String> = completed_tasks.into_iter()
                     .map(|task| task.address)
                     .collect();

                 let users_list: Vec<String> = user_set.into_iter().collect();
                 users_list
             }
             Err(e) => return get_error(format!("Error processing completed tasks: {}", e)),
         };

         (
             StatusCode::OK,
             Json(json!({ "users": users })).into_response(),
         )
     }
     ```

3. **Run Tests and Commit Changes**:

   - **Testing**:
     - **Functional Tests**:
       - Test with a valid `quest_id` that has tasks and participants.
       - Test with a `quest_id` that has tasks but no participants.
       - Test with an invalid `quest_id` to ensure appropriate error handling.
     - **Authentication Tests**:
       - Verify that unauthorized access is blocked by the `auth_middleware`.
     - **Edge Cases**:
       - Test with a quest that has no tasks to ensure the route handles this gracefully.

   - **Commit Changes**:

     ```bash
     git commit -m "Fix: Implement get_quest_users admin route"
     ```

## Required 📋

To keep our workflow smooth, please make sure you follow these guidelines:

- **Assignment**: Don't create a pull request if you weren’t assigned to this issue.
- **Timeframe**: Complete the task within **3 business days**.
- **Closing the Issue**: In your PR description, close the issue by writing `Close #[issue_id]`.
- **Review Process**:
  - Once you've submitted your PR, change the label to **"ready for review"**.
  - If changes are requested, address them and then update the label back to **"ready for review"** once done.
- **Testing**: Test your PR locally before pushing, and verify that tests and build are working after pushing.

Thank you for your contribution 🙏

⚠️ **WARNING**: Failure to follow the requirements above may result in being added to the OnlyDust blacklist, affecting your ability to receive future rewards.

### PR DESCRIPTION
**Pull request by MrRoudyk**
Changes Made:
Added get_quest_users Route
Added a public getter for `address` in `CompletedTaskDocument`
Allows external modules to access the private `address` field without exposing it directly

Closes #326

### CODE REVIEW

**Comment by MrRoudyk** [comment]
@Marchand-Nicolas can you take a look at this draft? I had to add lib.rs for testing. Not sure if that's right

File: config.template.toml [hunk 1]
```diff
@@ -123,3 +123,8 @@ contract = "0x0541b5dd5fae206ceccaf4eeb0642e4c04d456c5bc296eab047c9414bdad4f09"
 [quest_boost]
 private_key = "0xFFFFFFFFFFFF"
 update_interval = 600
+
+[get_quest_users]
```
Comment (Marchand-Nicolas) [inline config.template.toml]: No need to add that

File: Cargo.toml [hunk 2]
```diff
@@ -20,9 +20,9 @@ tower-http = { version = "0.4.0", features = ["cors"] }
 mongodb = "2.4.0"
 futures = "0.3.28"
 reqwest = { version = "0.12.0", features = ["rustls-tls", "json"] }
-reqwest-middleware = { version = "0.3", features = ["json"] }
-reqwest-retry = "0.5"
-reqwest-tracing = "0.5"
+reqwest-middleware = { version = "0.4", features = ["json"] }
```
Comment (Marchand-Nicolas) [inline Cargo.toml]: Please don't update the libs

File: src/config.rs [hunk 3]
```diff
@@ -112,6 +112,12 @@ pub_struct!(Clone, Deserialize;  QuestBoost{
     update_interval: u64,
 });
 
+pub_struct!(Clone, Deserialize;  GetQuestUsers{
+    test_user: String,
```
Comment (Marchand-Nicolas) [inline src/config.rs]: Please don't modify the config

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#[macro_use]
+pub mod utils;
```
Comment (Marchand-Nicolas) [inline src/lib.rs]: No need to have this file

File: src/main.rs [hunk 5]
```diff
@@ -1,43 +1,13 @@
-#[macro_use]
-mod utils;
-mod common;
-mod config;
```
Comment (Marchand-Nicolas) [inline src/main.rs]: Do no modify main.rs

File: src/models.rs [hunk 6]
```diff
@@ -127,7 +127,7 @@ pub struct ContractCall {
 
 #[derive(Debug, Serialize, Deserialize, Default)]
 pub struct QuestTaskDocument {
-    pub(crate) id: i32,
```
Comment (Marchand-Nicolas) [inline src/models.rs]: Why ?

File: tests/endpoints.rs [hunk 7]
```diff
@@ -1,29 +1,192 @@
...
+        let token = get_test_user_token().await;
+        let conf = config::load();
+        let quest_id = 2;
+        let endpoint = format!(
+            "http://0.0.0.0:{}/admin/quests/get_quest_users?quest_id={}",
+            &conf.server.port, quest_id
+        );
+        let client = reqwest::Client::new();
+        let response = client
+            .get(&endpoint)
```
Comment (Marchand-Nicolas) [inline tests/endpoints.rs]: No need to have tests for this route

**Review by Marchand-Nicolas — CHANGES_REQUESTED** [review]
Use the auth system we have already setup in other admin routes

```rs
    let res = verify_quest_auth(sub, &quests_collection, &(body.quest_id as i64)).await;
    if !res {
        return get_error("Error creating task".to_string());
    };
```

**Comment by MrRoudyk** [comment]
@Marchand-Nicolas thanks for the feedback. I'll make the corrections today and submit the PR, changing the label to "ready for review".

File: Cargo.toml (hunk 2 above)
Comment (MrRoudyk) [reply Cargo.toml]: I had to do this because the project wouldn't build until I updated the dependencies.
```
error[E0277]: the trait bound `TracingMiddleware<DefaultSpanBackend>: reqwest_middleware::Middleware` is not satisfied
--> src/endpoints/defi/rewards.rs:40:15
```
Now I saw the solution in telegram

File: tests/endpoints.rs (hunk 7 above)
Comment (MrRoudyk) [reply tests/endpoints.rs]: I probably misunderstood the tasks.
> Testing:
> Functional Tests:
> Test with a valid quest_id that has tasks and participants.
> Test with a quest_id that has tasks but no participants.
> Test with an invalid quest_id to ensure appropriate error handling.
> Authentication Tests:
> Verify that unauthorized access is blocked by the auth_middleware.
> Edge Cases:
> Test with a quest that has no tasks to ensure the route handles this gracefully.

But if tests are not needed, then how to implement these tasks?

**Review by Marchand-Nicolas — APPROVED** [review]
Working perfectly lgtm!
